### ISSUE TITLE
Installtion issue & how to identify tyre height using your model

### ISSUE THREAD

**Issue by Yogeshvasu** [bohundan/treadscan#2]
Initially, I'm impressed with the output your library produces.

However, I've encountered an issue while attempting to install your library. I consistently receive an error stating, "Could not build wheels for treadscan, which is required to install pyproject.toml-based projects." Can you provide insights into this issue and suggest potential solutions?

Additionally, I'm curious if your library supports image processing for detecting tire height. If so, could you please share some code examples demonstrating how to utilize this feature?

**Comment by bohundan** [comment]
Hello,

unfortunately, I can't assist you with the installation. I tested it on multiple Linux and Windows systems and didn't run into any issues, so I recommend just googling the errors that you get and looking for solutions online.

Just now I tried installing it by creating a fresh venv:
- `python3.9 -m venv ./venv3.9`
- `source ./venv/bin/activate`
- `pip install treadscan==1.0.5`

You don't need to install the `treadscan` package, you can just clone the repository and use the Python code. I'm guessing you have issues with the Cython compilation. The module written in Cython is responsible for the tread unwrapping part (its about 600x faster than raw Python). If you really want, you can dig for the older Python version in the git history.

To find the tire height, I would use the pretrained tire keypoint detecting model located in `RCNN_model/saved_model.pth`. Make sure to clone the repository using `git lfs` or just download this file manually.
Additionally, you can fine tune it using the tools provided in `RCNN_model` folder.

Below, I'm providing you with an example for detecting tire height:
```python
import cv2
import treadscan

# Substitute with the path to the downloaded pretrained model
segmentor = treadscan.SegmentorRCNN('saved_model.pth')

# Load your own image
image = cv2.imread('treadscan/extras/test_image.jpg', cv2.IMREAD_GRAYSCALE)

list_of_keypoints = segmentor.find_keypoints(image)

if len(list_of_keypoints) == 0:
    print('No tire keypoints found in image.')
else:
    keypoints = list_of_keypoints[0]
    # tire_model = treadscan.TireModel(image.shape)
    # tire_model.from_keypoints(*keypoints)
    t, b, r, s, w = keypoints

    rim_diameter = abs(t[1] - b[1])
    print(f'Rim diameter: {rim_diameter} px')
    tire_sidewall = abs(t[1] - s[1])
    print(f'Tire sidewall height: {tire_sidewall} px')

    tire_height = tire_sidewall * 2 + rim_diameter
    print(f'Tire height: {wheel_height} px')
``` 

Keep in mind this only gives you the tire dimensions in pixels. To translate to centimeters, you need have additional scene parameters from your camera calibration. You'll have to figure this out yourself :smile: good luck.

**Comment by Yogeshvasu** [comment]
Hi thanks for your solution.

 i ran your script but it shows No tire keypoints found in image.  i have executed in image where i have tire image 

but it shows No tire keypoints found in image. 

![34](https://github.com/bohundan/treadscan/assets/32808312/8c937465-0eb8-4164-90a6-38d441f72c5a)

**Comment by bohundan** [comment]
It's because the pretrained model has been trained on images of tires at a specific angle, all the training images are available in [this folder](https://github.com/bohundan/treadscan/tree/master/RCNN_model/dataset/images). There are always 5 locations for the 5 keypoints clearly visible.

Your image only has 4 visible keypoints. The 5th one for the tire width is obscured and the model does not know any other objects beside the 5 keypoint tire, so it fails. Remember this project was a part of my bachelor thesis and I did not possess the resources to obtain thousands of images AND label them accordingly. I just fine tuned the KeypointRCNN model to work for my use case and tested it on similar data. The total was less than 200 individual images.

If you want to, you can adapt the model for your use case by adding more/different training data. Take a look at [this notebook](https://github.com/bohundan/treadscan/blob/master/RCNN_model/KeypointRCNN_training.ipynb) to get an idea how to perform the training.

**Comment by Yogeshvasu** [comment]
Hi thanks for your model and solution , may i know for your dataset which annotation tool did you use

**Comment by bohundan** [comment]
I implemented my own using the opencv image view, because I wanted a preview of the constructed tire model and the unwrapped tread. The implementation can be found in the [extras](https://github.com/bohundan/treadscan/tree/master/extras) folder. There's also a short description in [this readme](https://github.com/bohundan/treadscan/blob/master/RCNN_model/README.md). But I'm sure there are other keypoint annotation tools out there.

**Comment by Yogeshvasu** [comment]
Thank you so much for clarifying the doubts.

Before proceeding with model training, I would like to address the
following points regarding the use of your model with a custom dataset. My
goal is to measure the height from the center of the wheel arc to the
contact point of the ground. To achieve this, I am annotating images within
the dataset by delineating specific regions and marking four key points:

   1. A - Center of the Wheel Arc
   2. R - Center of the Wheel Rim
   3. C - Center of the Wheel
   4. F - Ground Contact Point

Following annotation, I export the data as JSON and proceed with training
using your model. Please refer to the attached snapshot for clarity

Thanks

**Comment by bohundan** [comment]
Yes, this is the correct procedure. I followed the tutorial in [this article](https://medium.com/@alexppppp/how-to-train-a-custom-keypoint-detection-model-with-pytorch-d9af90e111da) to train my model. Best of luck in your project.

**Comment by Yogeshvasu** [comment]
Hello, thank you for confirmation.

After conducting tests on my images following training, it accurately identifies 5 points. However, in order to measure the height from the center of the wheel arc to the floor contact point, I attempted two methods:

Utilizing your approach.
Calculating the pixel variance between the top and bottom.
Both methods yield different values, and I'm encountering difficulties in establishing the pixel-to-CM or MM ratio. Could you kindly advise on which approach would best address my issue of measuring height in millimeters or centimeters?
**Your approach**
```
`import cv2
import treadscan

# Substitute with the path to the downloaded pretrained model
segmentor = treadscan.SegmentorRCNN('saved_model.pth')

# Load your own image
image = cv2.imread(r"C:\Program Files (x86)\Python39\V_3.9.8\soft\WPy64\notebooks\RCNN_model - Copy\dataset\images\0036.jpg", cv2.IMREAD_GRAYSCALE)

list_of_keypoints = segmentor.find_keypoints(image)

if len(list_of_keypoints) == 0:
    print('No tire keypoints found in image.')
else:
    keypoints = list_of_keypoints[0]
    tire_model = treadscan.TireModel(image.shape)
    tire_model.from_keypoints(*keypoints)
    t, b, r, s, w = keypoints

    rim_diameter = abs(b[1] - s[1])
    print(f'Rim diameter: {rim_diameter} px')
    tire_sidewall = abs(t[1] - w[1])
    print(f'Tire sidewall height: {tire_sidewall} px')

    tire_height = tire_sidewall * 2+ rim_diameter
    print(f'Tire height: {tire_height} px')`
```

My approach

```
import cv2
import numpy as np
import treadscan
# Substitute with the path to the downloaded pretrained model
segmentor = treadscan.SegmentorRCNN('saved_model.pth')
image = cv2.imread(r"C:\Program Files (x86)\Python39\V_3.9.8\soft\WPy64\notebooks\RCNN_model - Copy\dataset\images\0036.jpg", cv2.IMREAD_GRAYSCALE)
image = cv2.resize(image, (2420, 1815))
list_of_keypoints = segmentor.find_keypoints(image)
conversion_factor = 0.1
if len(list_of_keypoints) == 0:
    print('No tire keypoints found in image.')
else:
    keypoints = list_of_keypoints[0]
    t, b, r, s, w = keypoints
    points= np.array([
    [r[0], t[1]],   
    [r[0], b[1]],  
    [r[0], r[1]],  
    [r[0], s[1]],   
    [r[0], w[1]]
    ])
    cv2.rectangle(image, (r[0]-100, t[1]), (r[0]+100, w[1]), color=(255,0,0), thickness=2)
    cv2.imshow("lalala", image)
    cv2.waitKey(0) # 0==wait forever
    print(t[1])
    print(w[1])
```

![download](https://github.com/bohundan/treadscan/assets/32808312/13717944-5450-46fb-b7a6-7ed3d7a26931)

**Comment by bohundan** [comment]
As far as I know, optical measurements can be done if you have a defined scene, e.g., measuring size of a drilled hole in some industrial manufacturing setting, where the product is photographed at a constant distance everytime, and you know the exact camera configuration - the lens distortion, focal length...

Measuring any object without having some reference points is not easy, perhaps impossible, because as you said, you need to figure out the pixel-to-real-world ratio.

Also this is entirely off topic to the `treadscan` project so I'm not sure I'm the most qualified person you should be asking for help.

**Comment by Yogeshvasu** [comment]
Thanks for your suggestion. is your library treadscan can be integrated to android or IOS

**Comment by bohundan** [comment]
I don't understand why you're asking this here considering that is a general question, not a question specific to this repository. From a quick google search, yes, you can create mobile applications using Python, which the `treadscan` logic is written in.

From your previous comments, I assume you're more interested in using the RCNN model rather than the tread scanning utility. Since you've already played around with it, you could've also searched for how to port a PyTorch model to a mobile device. Here is the link anyway: https://pytorch.org/mobile/home/.

I'm going to close this issue because you seem to be asking me for advice on stuff that I'm not qualified to help you with, when there are much better places to search.
